### PR TITLE
repo-updater: make syncer & scheduler aware of uncloned gitserver repositories

### LINKED ISSUE #11091: cloning does not restart after gitservers are replaced
- **Sourcegraph version:** 3.16
- **Platform information:** GKE version 1.14

A large (140,000k) of repositories were waiting to be cloned after the gitserver stateful set had to be redeployed due to disk issues. However the gitservers, nor the repo-updater were acting on the queue on repos after redeployment. Only manually triggering the repos one by one would start a clone. The external source was then removed again. 

There was no log evidence to suggest a problem with either service. In the notifications in the UI, the ony message suggested that the repos were waiting to be cloned, with nothing happening. 

When removing the external source, and adding it back again, the repo-updater ran out of memory. This was increased, which then appeared to shift the load to postgres, which then had a memory limit increase and the sync completed with the repos from the external source no longer being in the queue. 

- **Expected behaviour**

This is potentially an extreme example, but not outside the realms of possibly for larger customers who need to increase storage on their cluster. At a minimum it should be possible to trigger a re-clone of the repos via the UI or src-cli. 

- **Actual behaviour**

When repos need to be cloned, there is no way to manually kick off this process, or evidence that something is happening when a large number of repos is in the queue

### PR DESCRIPTION
**Pull request by slimsag**
This is a fix for #11091 which we have learned has become a critical and blocking issue for a large customer ([details here](https://sourcegraph.slack.com/archives/CJX299FGE/p1592517101301600)) in the past day.

I chose to work after-hours to implement this as there is a heavy sense of urgency here. The implementation turned out to be trickier than I thought it would be: It was not 100% clear-cut to me where this logic should be structured and live, so it may make sense for someone with broader architect design goals in mind here to evaluate and adjust the implementation of this at a later date. This is not a completely unreasonable implementation, though, it just doesn't fit as seamlessly as I would like it to.

Prior to this change:

- repo-updater would instruct gitserver to update repositories if they were added, removed, or had their _metadata_ changed on the code host.
- gitserver would be responsible for keeping its own repositories up-to-date for Git updates by periodically running `git pull`.

However, it was possible for these two to get out of sync. For example from a completely clean slate when deploying a new Sourcegraph instance:

- repo-updater identifies thousands of new repositories on the code host not yet in the DB store.
- repo-updater instructs gitserver to clone the thousands of new repositories.
- gitserver's in-memory clone queue now has thousands of repositories and begins cloning.
- gitserver is restarted for any reason, its in-memory update queue is now gone.
- gitserver does not have the repositories cloned, it is completely unaware of them.
- repo-updater just naively assumes that gitserver has them cloned and does nothing to correct the situation.

This has particularly bad effects for users & admins:

1. **Searches are incredibly slow and timeout because e.g. half of your several thousand repos are not indexed.**
   - Why: zoekt can _discover_ that the repositories exist, but [when it attempts to fetch the tar archive it cannot because gitserver says the repository doesn't exist.](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7fcb1342431195e12e9d6e8242b9eb303540008c/-/blob/cmd/frontend/internal/httpapi/internal.go#L521))
2. **The admin UI shows "100,000 repositories enqueued for cloning..." yet the system never does anything**
   - Why: quite ironically, the status message code *does* compare with Gitserver how many repositories are uncloned - we just do absolutely nothing for that case in the actual code
3. **All logs, monitoring, and metrics just show "everything is fine" and "nothing is happening"**
   - Why: from the perspective of gitserver, everything _is cloned_ and from the naive perspective of repo-updater _everything is cloned_.
4. **There is no good workaround / way to correct this issue**
   - Why: You could delete *all* repositories and start over, but cloning takes a long time (and you better hope it doesn't happen again, as was the case with this customer)

Fixes #11091

### CODE REVIEW

File: cmd/repo-updater/repos/scheduler.go [hunk 1]
```diff
@@ -253,6 +253,11 @@ func (s *updateScheduler) UpdateFromDiff(diff Diff) {
 		s.upsert(r, false)
 	}
 
+	for _, r := range diff.NotCloned {
+		updated := s.updateQueue.enqueue(configuredRepo2FromRepo(r), priorityLow)
+		log15.Debug("scheduler.updateQueue.enqueued", "repo", r.Name, "updated", updated)
```
Comment (unknwon) [inline cmd/repo-updater/repos/scheduler.go:258]: What's bad about just call `s.upsert(r, true)`?

**Comment by tsenart** [comment]
@slimsag, @sourcegraph/cloud: This approach looks good to me and I extended it to also store the cloned status of each repo in the database. This will allow us to massively speed up the not cloned / cloned admin repo listing page.

**Comment by tsenart** [comment]
I have start a stacked PR that achieves the optimizations alluded to before: https://github.com/sourcegraph/sourcegraph/pull/11606

**Comment by mrnugget** [comment]
> This approach looks good to me and I extended it to also store the cloned status of each repo in the database.

Yes! Finally! 🙏 

>  This will allow us to massively speed up the not cloned / cloned admin repo listing page.

And also possibly the status messages: https://github.com/sourcegraph/sourcegraph/blob/716ceb75498932353008cdbb22e500d4221d15cb/cmd/repo-updater/repoupdater/server.go#L510-L514

Not sure though if that's what we want, because if the cloned status is updated only on sync then it might be less up-to-date than the status messages right now.

**Comment by keegancsmith** [comment]
About to review the code. Thank you for the detailed description and explanation for the current problem. It would be useful for your PR description to also say what it has done, rather than just relying on code :)

**Comment by ryanslade** [comment]
I'm happy to help out with this, especially since I'll be working on making `repo-updater` horizontally scalable (https://github.com/sourcegraph/sourcegraph/issues/11493)

**Comment by keegancsmith** [comment]
@slimsag I'm gonna timebox an hour now to try an alternative implementation.

**Comment by keegancsmith** [comment]
Took an hour to do :) I did however go grocery shopping in between so here is my PR now https://github.com/sourcegraph/sourcegraph/pull/11628

**Comment by keegancsmith** [comment]
I've merged in #11628 which is the alternate solution to this PR. I'm going to go ahead and close :) Thanks for the PR though and for lighting a fire under our ass to get this fixed :D I hope this solves the issue for the large instance you are testing.

**Pull request closed without merging**